components/brave_wallet/browser/siwe_message_parser.cc

```diff
@@ -0,0 +1,515 @@
+bool IsSubDelimChar(char c) {
+  switch (c) {
+    case '!':
+    case '$':
+    case '&':
+    case '\'':
+    case '(':
+    case ')':
+    case '*':
+    case '+':
+    case ',':
+    case ';':
+    case '=':
+      return true;
+  }
+  return false;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986/#section-3.3
+bool IsPChar(char c) {
+  if (IsUnreservedChar(c)) {
+    return true;
+  }
+  if (IsSubDelimChar(c)) {
+    return true;
+  }
+  switch (c) {
+    case ':':
+    case '[':
+    case ']':
+    case '@':
+      return true;
+  }
+  return false;
+}
+
+absl::optional<std::pair<std::string::size_type, std::string_view>>
+ExtractValue(std::string_view input,
+             std::string_view field,
+             bool expect_lf = true) {
+  auto field_end = field.length();
+  if (field_end == input.length()) {
+    return absl::nullopt;
+  }
+  auto field_pos = input.rfind(field);
+  if (field_pos == std::string::npos || field_pos != 0) {
+    return absl::nullopt;
+  }
+  std::string::size_type n;
+  if (expect_lf) {
+    n = input.find('\n', field_end);
+    if (n == field_end || n == std::string::npos) {
+      return absl::nullopt;
+    }
+  } else {
+    n = input.length();
+  }
+
+  return std::make_pair(n, input.substr(field_end, n - field_end));
+}
+
+absl::optional<std::pair<std::string::size_type, std::string_view>>
+ExtractOptionalValue(std::string_view input,
+                     std::string_view field,
+                     bool is_resources = false) {
+  auto effective_field = base::StrCat({"\n", field});
+  auto field_end = effective_field.length();
+  if (field_end == input.length()) {
+    return absl::nullopt;
+  }
+  auto field_pos = input.rfind(effective_field);
+  if (field_pos == std::string::npos) {
+    return std::make_pair(std::string::npos, input);
+  }
+  if (field_pos != 0) {
+    return absl::nullopt;
+  }
+  std::string::size_type n;
+  if (is_resources) {
+    n = input.length();
+  } else {
+    // Check if this is the last optional field
+    n = input.find('\n', field_end);
+    if (n == std::string::npos) {
+      return std::make_pair(input.length(), input.substr(field_end));
+    }
+    if (n == field_end) {
+      return absl::nullopt;
+    }
+  }
+
+  return std::make_pair(n, input.substr(field_end, n - field_end));
+}
+
+absl::optional<std::string_view> GetNextSegment(const std::string& message,
+                                                size_t index) {
+  if (index > message.size() - 1) {
+    return absl::nullopt;
+  }
+  return std::string_view(message.begin() + index, message.end());
+}
+
+}  // namespace
+
+SIWEMessageParser::SIWEMessageParser() = default;
+SIWEMessageParser::~SIWEMessageParser() = default;
+
+mojom::SIWEMessagePtr SIWEMessageParser::Parse(const std::string& message) {
+  state_ = State::kStart;
+  auto resultt = mojom::SIWEMessage::New();
+  size_t index = 0;
+  if (!ParseSchemeAndDomain(message, index, resultt->origin)) {
```

Comment:
`message` and `index` should be a string_view instance initially constructed from whole message.
Each of `Parse*` gets it by reference and 'extracts' a line(or more) from it, doing its job of parsing that line.
After each successful `Parse*` calls ends that string_view instance points to the next line.

I don't insist on this approach, but I think it would be a bit clearer.

Reply:
That is cleaner, fixed in 438de51dfbf6ced8bb1bbffbcb7077529e4bc8b8